File: Тренировки по алгоритмам 1.0/Лекция 1. «Сложность, тестирование, особые случаи»/J.py
def solve(a, b, c, d, e, f):
    if a == b == c == d == 0:
        if e == f == 0:
            return 5,
        else:
            return 0,

    if a == c == 0:
        if b == 0:
            if e != 0:
                return 0,
            return 4, f / d
        if d == 0:
            if f != 0:
                return 5,
            return 4, e / b
        if e / b != f / d:
            return 0,
        return 4, e / b

    if b == d == 0:
        if a == 0:
            if e != 0:
                return 0,
            return 3, f / c
        if c == 0:
            if f != 0:
                return 0,
            return 3, e / a
        if e / a != f / c:
            return 0,
        return 3, e / a

    if a == b == 0:
        if e != 0:
            return 0,
        return 1, -c / d, f / d

    if c == d == 0:
        if f != 0:
            return 0,
        return 1, -a / b, e / b

    if a == 0:
        return 2, (f - d * e / b) / c, e / b
    if b == 0:
        return 2, e / a, (f - c * e / a) / d
    if c == 0:
        return 2, (e - b * f / d) / a, f / d
    if d == 0:
        return 2, f / c, (e - a * f / c) / a

    if a / c == b / d:
        if e == f == 0:
            return 1, -a / b, 0
        if e == 0 or f == 0:
            return 0,
        if a / c != e / f:
            return 0,
        else:
            return 1, -a / b, e / b

    delta = a * d - c * b
    if delta == 0:
        return 0,
    x = (e * d - f * b) / delta
    y = (f * a - e * c) / delta
    return 2, x, y


assert solve(0, 0, 8, 4, 1, 5) == (0,)
assert solve(1, 2, 3, 4, 5, 6) == (2, -4, 4.5)

assert solve(1, 1, 2, 2, 0, 1) == (0,)
assert solve(1, 1, 2, 2, 0, 0) == (1, -1, 0)

assert solve(0, 0, 0, 0, 0, 0) == (5,)
assert solve(0, 0, 0, 0, 1, 0) == (0,)
assert solve(0, 0, 0, 2, 0, 5) == (4, 2.5)

assert solve(0, 2, 0, 4, 1, 2) == (4, 0.5)
assert solve(2, 0, 4, 0, 1, 2) == (3, 0.5)
assert solve(1, 0, 0, 1, 3, 3) == (2, 3, 3)
assert solve(0, 1, 1, 0, 3, 3) == (2, 3, 3)

assert solve(3, 4, 3, 4, 12, 18) == (0,)
assert solve(1, 1, 2, 2, 5, 7) == (0,)

assert solve(1, 1, 2, 2, 1, 2) == (1, -1, 1)
assert solve(0, 2, 0, 4, 1, 2) == (4, 0.5)

assert solve(1, 0, 0, 1, 3, 3) == (2, 3.00000, 3.00000)
assert solve(2, 2, 3, -3, 6, -3) == (2, 1.00000, 2.00000)
assert solve(1, 1, 1, -1, 3, -1) == (2, 1.00000, 2.00000)
assert solve(1, 1, 2, 2, 1, 2) == (1, -1.00000, 1.00000)
assert solve(1, 1, 1, 1, 1, 2) == (0,)
assert solve(0, 1, 0, 1, 2, 3) == (0,)
assert solve(0, 1, 0, 2, 2, 4) == (4, 2.00000)
assert solve(1, 0, 2, 0, 2, 4) == (3, 2.00000)
assert solve(0, 0, 0, 0, 0, 0) == (5,)
assert solve(0, 0, 0, 0, 1, 0) == (0,)
assert solve(0, 2, 0, 4, 1, 2) == (4, 0.50000)
assert solve(0, 2, 0, 4, 1, 2) == (4, 0.50000)
assert solve(0, 0, 2, 4, 0, 2) == (1, -0.50000, 0.50000)
assert solve(2, 4, 0, 0, 2, 0) == (1, -0.50000, 0.50000)
assert solve(2, 0, 3, 0, 2, 3) == (3, 1.00000)
assert solve(2, 2, 3, 0, 6, 3) == (2, 1.00000, 2.00000)
assert solve(1, 1, 1.5, 0, 3, 1.5) == (2, 1.00000, 2.00000)


def main():
    a = float(input())
    b = float(input())
    c = float(input())
    d = float(input())
    e = float(input())
    f = float(input())
    print(*solve(a, b, c, d, e, f))


if __name__ == '__main__':
    main()

File: Тренировки по алгоритмам 7.0/Лекция 3. Битовые операции, исправляющие коды Хэмминга, сжатие данных (алгоритм Лемпела-Зива)/J.cpp
#include <algorithm>
#include <bitset>
#include <cassert>
#include <cctype>
#include <cmath>
#include <cstddef>
#include <cstdint>
#include <fstream>
#include <iosfwd>
#include <iostream>
#include <limits>
#include <optional>
#include <queue>
#include <random>
#include <ranges>
#include <sstream>
#include <string>
#include <unordered_map>
#include <utility>
#include <vector>
#include <map>

using namespace std::string_literals;

class Arhiver {
protected:
	std::vector<unsigned char> packed;
	class BitStreamIn {
		Arhiver& owner;
		int block_index = 0;
		int bit_index = 0;
	public:

		explicit BitStreamIn(Arhiver& owner) : owner{ owner } {}
		BitStreamIn& operator >> (bool& out);
		operator bool () const;
	};

	class BitStreamOut {
		Arhiver& owner;
		unsigned char* block = nullptr;
		int bit_index = 0;
	public:

		explicit BitStreamOut(Arhiver& owner) : owner{ owner }, block{ owner.NextBlock() } {}
		BitStreamOut& operator << (bool out);
	};

	BitStreamIn InStream() { return BitStreamIn{ *this }; }
	BitStreamOut OutStream() { return BitStreamOut{ *this }; }
	unsigned char* NextBlock() { return &packed.emplace_back(); }
};

Arhiver::BitStreamOut& Arhiver::BitStreamOut::operator<<(bool out) {
	if (out) {
		*block |= 1 << bit_index;
	} else {
		*block &= ~(1 << bit_index);
	}
	if (++bit_index == 8) {
		block = owner.NextBlock();
		bit_index = 0;
	}
	return *this;
}

Arhiver::BitStreamIn& Arhiver::BitStreamIn::operator>>(bool& out) {
	if (block_index >= owner.packed.size()) {
		out = false;
	} else {
		out = owner.packed[block_index] & 1 << (bit_index++);
	}
	if (bit_index == 8) {
		++block_index;
		bit_index = 0;
	}
	return *this;
}

Arhiver::BitStreamIn::operator bool() const {
	return block_index < owner.packed.size();
}

class LWZArhiver : public Arhiver {
	static const char TERM = 'z' + 1;
	static const char EXTEND = 'z' + 2;
	static const char POS_WIDTH = 5;
	static const char DISPL = 'a';
	static const int CHAR_WIDTH = 5;
	struct WordInfo {
		int id = 0;
		char added{};
		int prev = 0;
	};

	std::vector<std::string> words;
	std::unordered_map<std::string, WordInfo> word_map;
	void Clear();

public:
	LWZArhiver() { Clear(); }
	auto Pack(const std::string& data) -> std::vector<unsigned char>;
	auto Unpack(std::vector<unsigned char> data) -> std::string;
};

void LWZArhiver::Clear() {
	words = { ""s };
	word_map = { {""s, { 0, 0, -1 }} };
	for (char c = 'a'; c <= 'z'; ++c) {
		words.push_back(""s + c);
		word_map[words.back()] = { int(words.size() - 1), c, 0 };
	}
}

auto LWZArhiver::Pack(const std::string& data) -> std::vector<unsigned char> {
	int pos_width = POS_WIDTH;
	int max_pos = 0;
	for (int i = 0; i < pos_width; ++i) {
		max_pos <<= 1;
		max_pos |= 1;
	}
	auto os = OutStream();
	std::string w, wc;
	const auto Write = [&os](int bit_cnt, int val) {
		while (bit_cnt--) {
			os << (val & 1);
			val >>= 1;
		}
	};
	for (char c : data) {
		wc = w + c;
		if (word_map.count(wc)) {
			w = wc;
			continue;
		}
		auto prev = word_map.find(w);
		WordInfo wi;
		wi.added = c;
		wi.id = words.size();
		wi.prev = prev->second.id;
		words.push_back(wc);
		word_map[wc] = wi;
		Write(pos_width, wi.prev);
		Write(CHAR_WIDTH, c - DISPL);
		if (wi.id == max_pos) {
			Write(pos_width, 0);
			Write(CHAR_WIDTH, EXTEND - DISPL);
			++pos_width;
			max_pos <<= 1;
			max_pos |= 1;
		}
		w.clear();
		wc.clear();
	}
	if (!wc.empty()) {
		Write(pos_width, word_map.at(wc).id);
	} else {
		Write(pos_width, 0);
	}
	Write(CHAR_WIDTH, TERM - DISPL);
	return move(packed);
}

auto LWZArhiver::Unpack(std::vector<unsigned char> data) -> std::string {
	packed = std::move(data);
	int pos_width = POS_WIDTH;
	int max_pos = 0;
	for (int i = 0; i < pos_width; ++i) {
		max_pos <<= 1;
		max_pos |= 1;
	}
	auto is = InStream();
	bool read_pos = true;
	std::ostringstream res;
	std::string buf;
	const auto Read = [&is](int bit_cnt) -> int {
		int res = 0;
		for (int i = 0; i < bit_cnt; ++i) {
			bool bit;
			is >> bit;
			res |= (bit << i);
		}
		return res;
	};
	while (is) {
		int id = Read(pos_width);
		char c = Read(CHAR_WIDTH) + DISPL;
		if (c == EXTEND) {
			++pos_width;
			max_pos <<= 1;
			max_pos |= 1;
			continue;
		}
		buf = words.at(id);
		if (c == TERM) {
			res << buf;
			break;
		}
		buf += c;
		res << buf;
		words.push_back(buf);
		buf.clear();
	}
	return res.str();
}

std::string generateRandomString(size_t length) {
	static const std::string chars = "abcdefghijklmnopqrstuvwxyz";
	static std::random_device rd;
	static std::mt19937 gen(rd());
	static std::uniform_int_distribution<> distrib(0, chars.size() - 1);

	std::string result;
	result.reserve(length);
	for (size_t i = 0; i < length; ++i) {
		result += chars[distrib(gen)];
	}
	return result;
}

void test() {
	while (true) {
		std::string s = generateRandomString(50000);
		//string s = "mobqgiqret";
		//ifstream file("input.txt");
		//file >> s;
		LWZArhiver a1;
		auto packed = a1.Pack(s);
		auto size = packed.size();
		std::cout << s.size() * 1. / packed.size() << std::endl;
		LWZArhiver a2;
		auto res = a2.Unpack(packed);
		if (res != s) {
			std::cout << s << std::endl;
			assert(res == s);
		}
	}
}

void LWZ() {
	LWZArhiver a;
	std::string cmd;
	std::cin >> cmd;
	if (cmd == "pack"s) {
		std::string data;
		std::cin >> data;
		auto packed = a.Pack(data);
		std::cout << packed.size() << std::endl;
		bool first = true;
		for (int c : packed) {
			if (first) {
				std::cout << c;
				first = false;
			} else {
				std::cout << ' ' << c;
			}
		}
		std::cout << std::endl;
	}
	if (cmd == "unpack"s) {
		int byte_count = 0;
		std::cin >> byte_count;
		std::vector<unsigned char> packed(byte_count);
		for (int i = 0; i < byte_count; ++i) {
			int byte = 0;
			std::cin >> byte;
			packed.at(i) = byte;
		}
		auto unpacked = a.Unpack(std::move(packed));
		std::cout << unpacked << std::endl;
	}
}

int main() {
	LWZ();
	return 0;
}
